src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.MoveFileCodeAction.cs

```diff
@@ -0,0 +1,162 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.SyncNamespace
+{
+    internal abstract partial class AbstractSyncNamespaceCodeRefactoringProvider<TService, TNamespaceDeclarationSyntax, TCompilationUnitSyntax>
+        where TService : AbstractSyncNamespaceCodeRefactoringProvider<TService, TNamespaceDeclarationSyntax, TCompilationUnitSyntax>
+        where TNamespaceDeclarationSyntax : SyntaxNode
+        where TCompilationUnitSyntax : SyntaxNode 
+    {
+        private class MoveFileCodeAction : CodeAction
+        {
+            private readonly Document _document;
+            private readonly ImmutableArray<string> _newfolders;
+
+            public override string Title
+                => _newfolders.Length > 0 
+                ? string.Format(FeaturesResources.Move_file_to_0_folder, string.Join(".", _newfolders)) 
+                : FeaturesResources.Move_file_to_project_root_folder;
+
+
+            public MoveFileCodeAction(Document document, ImmutableArray<string> newFolders)
+            {
+                _document = document;
+                _newfolders = newFolders;
+            }
+
+            protected override async Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(CancellationToken cancellationToken)
+                => await MoveFileToMatchNamespaceAsync(cancellationToken).ConfigureAwait(false);
+
+            private async Task<ImmutableArray<CodeActionOperation>> MoveFileToMatchNamespaceAsync(CancellationToken cancellationToken)
+            {
+                var newSolution = _document.Project.Solution.RemoveDocument(_document.Id);
+                var text = await _document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+                var newDocumentId = DocumentId.CreateNewId(_document.Project.Id, _document.Name);
+                newSolution = newSolution.AddDocument(newDocumentId, _document.Name, text: string.Empty, folders: _newfolders);
+                newSolution = newSolution.WithDocumentText(newDocumentId, text, PreservationMode.PreserveIdentity);
+
+                return ImmutableArray.Create<CodeActionOperation>(
+                    new ApplyChangesOperation(newSolution),
+                    new OpenDocumentOperation(newDocumentId, activateIfAlreadyOpen: true));
+            }
+            
+            public static ImmutableArray<MoveFileCodeAction> Create(State state)
+            {
+                var document = state.Document;
+
+                var parts = state.RelativeDeclaredNamespace.Split(new[] { '.' }).ToImmutableArray();
+                if (parts.Any(s => s.IndexOfAny(Path.GetInvalidPathChars()) >= 0))
```

Comment:
when would this happen?

Reply:
This check should be removed now. When there's error in the name of namespace declaration, it might end up containing invalid characters. I later decided to just check for error in the node instead.

Here's an example, where the name of namespace contains newline.
```cs
namespace
{
    class C1 { }
}
```